python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -358,6 +378,64 @@ def processResults(self, science, matchedTemplate, difference, sources, table,
 
         return measurementResults
 
+    def _deblend(self, difference, positiveFootprints, negativeFootprints, idFactory):
+        """Deblend the positive and negative footprints and return a catalog
+        containing just the children, and the deblended footprints.
+
+        Parameters
+        ----------
+        difference : `lsst.afw.image.Exposure`
+            Result of subtracting template from the science image.
+        positiveFootprints, negativeFootprints : `lsst.afw.detection.FootprintSet`
+            Positive and negative polarity footprints measured on
+            ``difference`` to be deblended separately.
+        idFactory : `lsst.afw.table.IdFactory`
+            Generator object to assign ids to detected sources.
+
+        Returns
+        -------
+        sources : `lsst.afw.table.SourceCatalog`
+            Positive and negative deblended children.
+        positives, negatives : `lsst.afw.detection.FootprintSet`
+            Deblended positive and negative polarity footprints measured on
+            ``difference``.
+        """
+        def makeFootprints(sources):
+            footprints = afwDetection.FootprintSet(difference.getBBox())
+            footprints.setFootprints([src.getFootprint() for src in sources])
+            return footprints
+
+        table = afwTable.SourceTable.make(self.schema, idFactory)
+        table.setMetadata(self.algMetadata)
+
+        def deblend(footprints):
+            """Deblend a positive or negative footprint set,
+            and return the deblended children.
+            """
+            sources = afwTable.SourceCatalog(table)
+            footprints.makeSources(sources)
+            self.deblend.run(exposure=difference, sources=sources)
+            self.setPrimaryFlags.run(sources)
+            children = sources["detect_isDeblendedSource"] == 1
+            sources = sources[children].copy(deep=True)
+            # Clear parents, so that measurement plugins behave correctly.
+            sources['parent'] = 0
+            return sources.copy(deep=True)
+
+        positives = deblend(positiveFootprints)
+        negatives = deblend(negativeFootprints)
+
+        # TODO: id generation here might be a problem; we are reusing the same
+        # id generator, which might just keep adding to the next id, instead
+        # of starting from scratch for the final catalog.
+        # TODO solution?: just use idFactory post-deblending, not in initial detection?
+        # TODO: also do that in CalibrateImage?
+        sources = afwTable.SourceCatalog(table)
```

Comment:
After some testing on real data, yes, the idFactory needs to be used post deblending. In my tests I put it after the `doMerge` block in processResults, and that fixed the duplicate diaObject bug I was seeing when running on multiple visits. My VERY hacky implementation there just forcibly set the diaSourceId via `initialDiaSources.getIdKey()`, which did work but I imagine there is a better way.

Reply:
Thanks for checking this. Can you give me a recipe for how you checked it on "real data"? I'd like to see what it should be doing, so I can make sure there's a test that checks that.

Reply:
What I did:
I ran the original code through `ap_verify` with the `ap_verify_ci_hits2015` dataset, and it crashed in `diaPipe` because of duplicate diaObjects. I re-ran with `main`, and compared the diaSource tables in a notebook, and noticed that the diaSource Ids on `main` were numbers like `88351793720852791` while on the branch they were ~`200`.
I checked that the diaSource Ids were correct with ipdb, and found that they were up until the `makeFootprints` call in the return statement of `_deblend`, after which they lost their ids. I couldn't find a way to set the id in the call signature for afw's `setFootprints`, so I just wrote the ids of the final catalog by force with the `idGenerator`.

A way to test this would be to call the `idGenerator` once before passing it in to `detectAndMeasure.run`, and then a second time after it completes. All of the diaSource Ids in the output catalog should have values greater than the first call and less than the second.